sycl/doc/extensions/SubGroup/SYCL_INTEL_sub_group.asciidoc

```diff
@@ -0,0 +1,251 @@
+= SYCL_INTEL_sub_group
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.  GitHub does not render image icons.
+
+This document describes an extension which introduces a new +sub_group+ class representing an implementation-defined grouping of work-items in a work-group.
+
+== Name Strings
+
++SYCL_INTEL_sub_group+
+
+== Notice
+
+Copyright (c) 2019-2020 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contact
+John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com)
+
+== Dependencies
+
+This extension is written against the SYCL 1.2.1 specification, Revision 6.
+
+== Overview
+
+A sub-group represents an implementation-defined grouping of work-items in a work-group. The work-items within a sub-group can communicate and synchronize independently of work-items in other sub-groups, and sub-groups are therefore commonly mapped to SIMD hardware where it exists.
```

Comment:
In OpenCL, subgroups are always 1-dimensional, do we want to clearly state the same for SYCL?

> Within a work-group work-items may be divided into sub-groups. The mapping of work-items to
sub-groups is implementation-defined and may be queried at runtime. While sub-groups may be
used in multi-dimensional work-groups, each sub-group is 1-dimensional and any given workitem may query which sub-group it is a member of.

However, even if this spelling theoretically allows 2d-subgroups, methods of the `sub_group` class below suggest that they are 1-dimensional.

BTW, OpenCL also provides some guarantees on work-items to sub-groups mapping during kernel execution, between dispatches of a given kernel and query operations:

> Work items are mapped into sub-groups through a combination of compile-time decisions and
the parameters of the dispatch. The mapping to sub-groups is invariant for the duration of a
kernel’s execution, across dispatches of a given kernel with the same work-group dimensions,
between dispatches and query operations consistent with the dispatch parameterization, and from
one work-group to another within the dispatch (excluding the trailing edge work-groups in the
presence of non-uniform work-group sizes). In addition, all sub-groups within a work-group will
be the same size, apart from the sub-group with the maximum index which may be smaller if the
size of the work-group is not evenly divisible by the size of the sub-groups.
> 
> In the degenerate case, a single sub-group must be supported for each work-group. In this
situation all sub-group scope functions are equivalent to their work-group level equivalents.

Reply:
Updated to partially address this.  We need to consider the degenerate case more, as I'm not sure we necessarily need to follow OpenCL here.